[DOCS] Advanced Tutorials for Daily-Quests and NPC-Images (#2903)
BetonQuest/BetonQuest: docs/Tutorials/Advanced-Tutorials/Daily-Quests/Delays.md

---
icon: material/folder-play-outline
tags:

>   - Delay

Reviewer: Yeah I know there are many ways to receive a specific thing but wouldn't it be better for perfomance and for easier understanding/writing to schedules? Correct me if I am wrong

  - Daily Quest
---

The next tutorial focuses on creating a delay that resets our daily quest after 24 hours. This will ensure players can 
complete another quest after a full day has passed. Follow along to implement this feature seamlessly.

<div class="grid" markdown>
!!! example "Related Docs"
    * [Events List](../../../Documentation/Scripting/Building-Blocks/Events-List.md)
    * [Objectives List](../../../Documentation/Scripting/Building-Blocks/Objectives-List.md)
</div>

> ## 1. Start a delay event

Reviewer: ```suggestion
## 1. Start a delay objective
```

Open the `events.yml` from your QuestPackage and add the following content to it

``` YAML title="events.yml" linenums="1"
events: # (1)!

>   start_reset_timer: "objective add reset_timer"

Reviewer: ```suggestion
  startResetTimer: "objective add resetTimer"
```

```

1. All events must be defined in an `events` section.

Afterwards we add a new objective into our `objectives.yml` where we start the delay.

``` YAML title="events.yml" hl_lines="11" linenums="1"
events:

>   pickdaily: "pickrandom 50%daily1,50%daily2"

Reviewer: ```suggestion
  pickdaily: "pickrandom 50%daily_1,50%daily_2"
```

>   daily1: 'objective add mine_diamonds'
>   daily2: 'objective add mine_iron'

Reviewer: ```suggestion
  daily_1: 'objective add mineDiamonds'
  daily_2: 'objective add mineIron'
```

>   dailyComplete: 'folder addQuestDoneTag,NotifyPlayer'

Reviewer: Typo
```suggestion
  dailyComplete: 'folder addQuestDoneTag,notifyPlayer'
```

  notifyPlayer: 'notify You mined enough!\nReturn to the Blacksmith! io:Title sound:firework_rocket'
  addQuestDoneTag: 'tag add questDone'
  

>   start_reset_timer: "objective add reset_timer"

Reviewer: ```suggestion
  startResetTimer: "objective add resetTimer"
```

```

Before we move on to our `objectives.yml`, let's add a folder event and several smaller events to remove all tags and
objectives that the player might have. In this example, we assume the player has either a `startedQuest` or a `questDone` 
tag. We will remove both and all possible objectives in case a player did not finish the quest within 24 hours.

``` YAML title="events.yml" hl_lines="5 7 12 14-15 19" linenums="1"

> events:
>   pickdaily: "pickrandom 50%daily1,50%daily2"
>   
>   daily1: 'objective add mine_diamonds'
>   daily1_r: 'objective remove mine_diamonds'
>   daily2: 'objective add mine_iron'
>   daily2_r: 'objective remove mine_iron'
>   
>   dailyComplete: 'folder addQuestDoneTag,NotifyPlayer'
>   notifyPlayer: 'notify You mined enough!\nReturn to the Blacksmith! io:Title sound:firework_rocket'
>   addQuestDoneTag: 'tag add questDone'
>   addQuestDoneTag_r: 'tag add questDone'
>   
>   addQuestStartedTag: 'tag remove questStarted' # (1)!
>   addQuestStartedTag_r: 'tag remove questStarted'
>   
>   start_reset_timer: 'objective add reset_timer'
>   
>   folderReset: 'folder daily1_r,daily2_r,addQuestStartedTag_r,addQuestDoneTag_r'

Reviewer: As I wrote above I would use schedules :D But this can still be used for the schedules!
And here are some mistakes I guess that needs to be looked at
```suggestion
events:
  pickdaily: "pickrandom 50%daily_1,50%daily_2"
  
  addDaily_1: 'objective add mineDiamonds'
  removeDaily_1: 'objective remove mineDiamonds'
  addDaily_2: 'objective add mineIron'
  removeDaily_2: 'objective remove mineIron'
  
  dailyComplete: 'folder addQuestDoneTag,notifyPlayer'
  notifyPlayer: 'notify You mined enough!\nReturn to the Blacksmith! io:Title sound:firework_rocket'
  addQuestDoneTag: 'tag add questDone'
  removeQuestDoneTag: 'tag remove questDone'
  
  addQuestStartedTag: 'tag add questStarted' # (1)!
  removeQuestStartedTag: 'tag remove questStarted'
  
  startResetTimer: 'objective add resetTimer'
  
  folderReset: 'folder removeDaily_1,removeDaily_2,removeQuestStartedTag,removeQuestDoneTag'
```

```

> 1. It works the same if you don't use points instead of tags.

Reviewer: I dont get it too. ^^ What do you mean with that? Should the sentence be like: "1. It works the same if you use points instead of tags."
But yeah its a bit random..

So what do we see here?

> * `dailyX_r`  is reversing the dailyX event, essentially removing the objective.
> * `addQuestXXXXTag_r` is reversing the tag event, removing the tag from the player

Reviewer: ```suggestion
* `removeDaily_X`  is reversing the daily_X event, essentially removing the objective.
* `removeQuestXXXXTag` is reversing the tag event, removing the tag from the player
```

* `folderReset` will be used as folder event that runs once 24 hours are over

## 2. Create the delay objective

In the next step we will create our delay objective to reset the quest 24 hours after the player finished the daily
quest on his part.

``` YAML title="objectives.yml" hl_lines="4" linenums="1"

> objectives:
>   mine_diamonds: "block DIAMOND_ORE -16 events:dailyComplete" 
>   mine_iron: "block IRON_ORE -16 events:dailyComplete"
>   reset_timer: "delay 86400 seconds events:folderReset"

Reviewer: ```suggestion
objectives:
  mineDiamonds: "block DIAMOND_ORE -16 events:dailyComplete" 
  mineIron: "block IRON_ORE -16 events:dailyComplete"
  resetTimer: "delay 86400 seconds events:folderReset"
```

```

This creates a timer that runs 24 hours and fires the folderReset event-folder once complete. You can also use schedules

> to reset the quests for all players at exactly midnight, but this would screw players that just recently started their

Reviewer: No its not if you have the reset schedule bind to a questDoneTag condition

quest.

Your Quest should now follow this structure:

!!! info "Daily-Quest Cycle"
    ``` mermaid
    graph LR
    X{Player accepts Quest} --> A
    X --> B

>     A[Add StartedTag] --> D[Player executes objective];
>     B[Pickrandom Task] --> D[Player executes objective];

Reviewer: ```suggestion
    A[Add StartedTag] --> D[Player completes objective];
    B[Pickrandom Task] --> D[Player completes objective];
```
